wagtail/contrib/wagtailapi/api.py

```diff
@@ -0,0 +1,535 @@
+
+    def json_response(self, data, response_cls=HttpResponse):
+        """
+        This takes a JSON-serialisable thing and builds a HTTP response
+        from it
+        """
+        return response_cls(
+            json.dumps(data, indent=4, cls=WagtailAPIJSONEncoder),
+            content_type='application/json'
+        )
+
+    def api_view(self, view):
+        """
+        This is a decorator that is applied to all API views.
+
+        It's only job currently is to catch Http404 and BadRequestError
+        exceptions and convert them into nicer error messages for the user.
+        """
+        @wraps(view)
+        def wrapper(*args, **kwargs):
+            # Catch exceptions and format them as JSON documents
+            try:
+                return view(*args, **kwargs)
+            except Http404 as e:
+                return self.json_response({
+                    'message': str(e)
+                }, response_cls=HttpResponseNotFound)
+            except self.BadRequestError as e:
+                return self.json_response({
+                    'message': str(e)
+                }, response_cls=HttpResponseBadRequest)
+
+        return wrapper
+
+    def get_urlpatterns(self):
+        """
+        This returns a list of URL patterns for the endpoint
+        """
+        return [
+            url(r'^$', self.api_view(self.listing_view), name='listing'),
+            url(r'^(\d+)/$', self.api_view(self.detail_view), name='detail'),
+        ]
+
+
+class PagesAPIEndpoint(BaseAPIEndpoint):
+    known_query_parameters = BaseAPIEndpoint.known_query_parameters + (
+        'type',
+        'child_of',
+    )
+
+    def get_queryset(self, request, model=Page):
+        # Get live pages that are not in a private section
+        queryset = model.objects.public().live()
+
+        # Filter by site
+        queryset = queryset.descendant_of(request.site.root_page, inclusive=True)
+
+        return queryset
+
+    def get_api_fields(self, model):
+        api_fields = ['title']
+        api_fields.extend(super(PagesAPIEndpoint, self).get_api_fields(model))
+        return api_fields
+
+    def serialize_object_metadata(self, request, page, show_details=False):
+        data = super(PagesAPIEndpoint, self).serialize_object_metadata(request, page, show_details=show_details)
+
+        # Add type
+        data['type'] = page.specific_class._meta.app_label + '.' + page.specific_class.__name__
+
+        # Add parent id
+        if show_details:
+            parent_id = page.get_parent().id
+
+            # Make sure the parent is visible in the API
+            if self.get_queryset(request).filter(id=parent_id).exists():
+                data['parent'] = page.get_parent().id
+
+        return data
+
+    def get_model(self, request):
+        if 'type' not in request.GET:
+            return Page
+
+        model_name = request.GET['type']
+        try:
+            return resolve_model_string(model_name)
```

Comment:
This should enforce that models are subclasses of `wagtailcore.Page`, otherwise you could potentially search for administrator accounts in the users table, for example.